sklearn/cluster/tests/test_optics.py

```diff
@@ -93,50 +92,34 @@ def test_extract_xi():
     C6 = [5, 6] + .2 * rng.randn(n_points_per_cluster, 2)
 
     X = np.vstack((C1, C2, C3, C4, C5, np.array([[100, 100]]), C6))
-    expected_labels = np.r_[[2] * 5, [0] * 5, [1] * 5, [3] * 5, [1] * 5,
+    expected_labels = np.r_[[0] * 5, [1] * 5, [2] * 5, [3] * 5, [2] * 5,
                             -1, [4] * 5]
-    X, expected_labels = shuffle(X, expected_labels, random_state=rng)
-
-    clust = OPTICS(min_samples=3, min_cluster_size=2,
-                   max_eps=np.inf, cluster_method='xi',
-                   xi=0.4).fit(X)
-    assert_array_equal(clust.labels_, expected_labels)
+    X, expected_labels = shuffle(X, expected_labels, random_state=0)
+    clust = OPTICS(min_samples=3, min_cluster_size=3,
+                   max_eps=20, cluster_method='xi').fit(X)
+    assert np.isclose(v_measure_score(clust.labels_, expected_labels), 1)
+    assert np.array_equal(np.where(clust.labels_ == -1)[0],
+                          np.where(expected_labels == -1)[0])
 
     X = np.vstack((C1, C2, C3, C4, C5, np.array([[100, 100]] * 2), C6))
-    expected_labels = np.r_[[1] * 5, [3] * 5, [2] * 5, [0] * 5, [2] * 5,
+    expected_labels = np.r_[[0] * 5, [1] * 5, [2] * 5, [3] * 5, [2] * 5,
                             -1, -1, [4] * 5]
-    X, expected_labels = shuffle(X, expected_labels, random_state=rng)
-
     clust = OPTICS(min_samples=3, min_cluster_size=3,
-                   max_eps=np.inf, cluster_method='xi',
-                   xi=0.1).fit(X)
-    # this may fail if the predecessor correction is not at work!
-    assert_array_equal(clust.labels_, expected_labels)
+                   max_eps=20, cluster_method='xi').fit(X)
+    assert np.isclose(v_measure_score(clust.labels_, expected_labels), 1)
+    assert np.array_equal(np.where(clust.labels_ == -1)[0],
+                          np.where(expected_labels == -1)[0])
 
     C1 = [[0, 0], [0, 0.1], [0, -.1], [0.1, 0]]
     C2 = [[10, 10], [10, 9], [10, 11], [9, 10]]
     C3 = [[100, 100], [100, 90], [100, 110], [90, 100]]
     X = np.vstack((C1, C2, C3))
     expected_labels = np.r_[[0] * 4, [1] * 4, [2] * 4]
-    X, expected_labels = shuffle(X, expected_labels, random_state=rng)
-
     clust = OPTICS(min_samples=2, min_cluster_size=2,
-                   max_eps=np.inf, cluster_method='xi',
-                   xi=0.04).fit(X)
-    assert_array_equal(clust.labels_, expected_labels)
-
-
-def test_cluster_hierarchy_():
```

Comment:
I actually think this is a test we need to have. We have nothing else testing the hierarchy.

Reply:
Yes, but this test is not reasonable IMO. Some points in C2 will also satisfy -2<=x<=2 & -2<=y<=2 so I can't understand why the expected clusters should be np.array([[0, 99], [0, 199]]).

Reply:
Actually the test easily fails if you change a random state when generating the dataset (you don't set a random state in the test so it's difficult to reproduce)